subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/dependency_verification.adoc

```diff
@@ -0,0 +1,900 @@
+
+[WARNING]
+====
+Signatures _can_ also used to assess the integrity of a dependency similarly to checksums.
+Signatures are signatures of the _hash_ of artifacts, not artifacts themselves.
+This means that if the signature is done on an _unsafe hash_ (even SHA1), then you're not correctly assessing the _integrity_ of a file.
+For this reason, if you care about both, you need to add both signatures _and_ checksums to your verification metadata.
+====
+
+However:
+
+- Gradle only supports verification of signatures published on remote repositories as ASCII-armored PGP files
+- Not all artifacts are published with signatures
+- A good signature doesn't mean that the signatory was legit
+
+As a consequence, signature verification will often be used alongside checksum verification.
+
+[NOTE]
+.About expired keys
+====
+It's very common to find artifacts which are signed with an expired key.
+This is not a problem for _verification_: key expiry is mostly used to avoid signing with a stolen key.
+If an artifact was signed before expiry, it's still valid.
+====
+
+=== Enabling signature verification
+
+Because verifying signatures is more expensive (both I/O and CPU wise) and harder to check manually, it's not enabled by default.
+
+Enabling it requires you to change the configuration option in the `verification-metadata.xml` file:
+
+[source,xml]
+----
+<?xml version="1.0" encoding="UTF-8"?>
+<verification-metadata>
+   <configuration>
+      <verify-signatures>true</verify-signatures>
+   </configuration>
+</verification-metadata>
+----
+
+[[sec:understanding-signature-verification]]
+== Understanding signature verification
+
+Once signature verification is enabled, for each artifact, Gradle will:
+
+* try to download the corresponding `.asc` file
+* if it's present
+** automatically download the keys required to perform verification of the signature
+** verify the artifact using the downloaded public keys
+** if signature verification passes, perform additional requested checksum verification
+* if it's absent, fallback to checksum verification
+
+That is to say that Gradle verification mechanism is much stronger if signature verification is enabled than just with checksum verification.
+In particular:
+
+- if an artifact is signed with multiple keys, all of them must pass validation or the build will fail
+- if an artifact passes verification, any additional checksum configured for the artifact _will also be checked_
+
+However, it's not because an artifact passes signature verification that you can trust it: you need to _trust the keys_.
+
+In practice, it means you need to list the keys that you trust for each artifact, which is done by adding a `pgp` entry instead of a `sha1` for example:
+
+[source,xml]
+----
+<component group="com.github.javaparser" name="javaparser-core" version="3.6.11">
+   <artifact name="javaparser-core-3.6.11.jar">
+      <pgp value="379ce192d401ab61"/>
```

Comment:
Is full fingerprint allowed here? I know that long key IDs are internally used everywhere but OpenPGP has since been extended with [Issuer Fingerprint subpacket](https://dev.gnupg.org/T4046) that can store entire fingerprint (I guess Java libraries may not expose it yet but it's already used and inserted by GnuPG).

Short key IDs (32bit) [are insecure](https://evil32.com/) but [there are collisions for 64 bit long key IDs too](https://nullprogram.com/blog/2019/07/22/).

Reply:
That is a good question. AFAIK, BouncyCastle implementation is tied to 64bit ids.

>there are collisions for 64 bit long key IDs too.

The collisions are based on generating a birthday paradox (or something like that).
In other words, there are no known attacks to generate a collision for a given input key.

However, it would probably make sense to support (or even always require) full fingerprint if it is not hard from implementation point of view.